src/app/core/device.ts

```diff
@@ -0,0 +1,1036 @@
+     * @memberof Device
+     * @returns {void}
+     */
+    addComponent(component: Component): void {
+        if (component instanceof Component) {
+            this.__components.push(component);
+        } else {
+            throw new Error("Tried to add a component that isn't a component to the device");
+        }
+    }
+
+    /**
+     * Removes a component from the device
+     * @param {Component} component Component to remove from the device
+     * @memberof Device
+     * @returns {void}
+     */
+    removeComponent(component: Component): void {
+        //Remove the component from the map
+        let trydelete;
+        let componentid = component.getID();
+        let connectiontorefresh = null;
+
+        //Remove component from connections
+        for (let i in this.__connections) {
+            let connection = this.__connections[i];
+            try {
+                trydelete = connection.tryDeleteConnectionTarget(componentid);
+                if (trydelete) {
+                    console.log("Removed Component from Connection : ", connection.getID());
+                }
+            } catch (e) {
+                console.error(e);
+            }
+        }
+        //Check if the valve map has the component
+        if (this.__valveMap.has(componentid)) {
+            connectiontorefresh = this.getConnectionByID(this.__valveMap.get(componentid));
+            this.__valveMap.delete(componentid);
+        }
+
+        let i = this.__components.indexOf(component);
+        if (i != -1) {
+            this.__components.splice(i, 1);
+        }
+
+        if (connectiontorefresh) {
+            Registry.viewManager.updatesConnectionRender(connectiontorefresh);
+        }
+    }
+
+    /**
+     * Returns the list of components from the device
+     * @return {Array<Component>} Array with the components of the device
+     * @memberof Device
+     */
+    get components(): Array<Component> {
+        return this.__components;
+    }
+
+    /**
+     * Sets the name of the device
+     * @param {string} name Name of the device
+     * @memberof Device
+     */
+    set name(name: string | String) {
+        this.__name = new StringValue(name);
+        this.updateView();
+    }
+
+    /**
+     * Updates the parameter
+     * @param {string} key Key to identigy the parameter
+     * @param {} value Parameter's value
+     * @memberof Device
+     * @returns {void}
+     */
+    updateParameter(key: string, value: any) {
+        this.__params.updateParameter(key, value);
+        this.updateView();
+    }
+
+    /**
+     * Sort the layers such that they are ordered from lowest to highest z_offset
+     * @returns {void}
+     * @memberof Device
+     */
+    sortLayers(): number {
+        //Does this need to be set up as a function?
+        this.__layers.sort(function(a, b) {
+            return a.params.getValue("z_offset") - b.params.getValue("z_offset");
+        });
```

Comment:
Where are 'a' and 'b' coming from? Is there any reason this needs to be a function within a function?

Reply:
Well this is a prototype for a function, you can use `Feature` for `a`,`b` for this because it tries to sort those layers